test/core/gc.spec.js

```diff
@@ -0,0 +1,279 @@
+/* eslint max-nested-callbacks: ["error", 8] */
+/* eslint-env mocha */
+'use strict'
+
+const chai = require('chai')
+const dirtyChai = require('dirty-chai')
+const expect = chai.expect
+chai.use(dirtyChai)
+
+const IPFSFactory = require('ipfsd-ctl')
+const pEvent = require('p-event')
+const env = require('ipfs-utils/src/env')
+const IPFS = require('../../src/core')
+
+// We need to detect when a readLock or writeLock is requested for the tests
+// so we override the Mutex class to emit an event
+const EventEmitter = require('events')
+const Mutex = require('../../src/utils/mutex')
+
+class MutexEmitter extends Mutex {
+  constructor (repoOwner) {
+    super(repoOwner)
+    this.emitter = new EventEmitter()
+  }
+
+  readLock (lockedFn, cb) {
+    this.emitter.emit('readLock request')
+    return super.readLock(lockedFn, cb)
+  }
+
+  writeLock (lockedFn, cb) {
+    this.emitter.emit('writeLock request')
+    return super.writeLock(lockedFn, cb)
+  }
+}
+
+describe('gc', function () {
```

Comment:
A couple of questions:
- What is the largest repo size that gc has been tested with? 
- How do these tests compare to go-ipfs gc tests?

Reply:
I'm asking this as I expect people to use GC when their repo gets of a significant size (10GB +++) and not when it is small.

Reply:
Good point 👍

I have been testing locally with repos a few hundred MB in size:
```
$ du -h ~/.jsipfs | tail -1
621M	/Users/dirk/.jsipfs

$ time js-ipfs repo gc
real	0m4.141s
user	0m0.858s
sys	0m0.174s
```

I haven't performed stress testing with repos of the size you're suggesting.

I had a look at go-ipfs, it seems like there are [some sharness tests](https://github.com/ipfs/go-ipfs/blob/777d306f6e66e31a05f43a337bce272050407386/test/sharness/t0082-repo-gc-auto.sh) for GC, but I didn't see any stress tests.

Reply:
I just tried with an 11GB repo and it ran in a few seconds. I will add a sharness test for stress-testing

Reply:
11GB repo and how large of a GC? half of it?

Reply:
GC removes all blocks that are not pinned (same as go-ipfs). In this case I added 11GB of unpinned files and then GCed them.

Reply:
I added a [sharness test for GC](https://github.com/ipfs/js-ipfs/pull/2022/commits/3a4912f84d92d187df8314e12cbd3ebb1d009610), in which the number of files and their size can be easily adjusted. It seems like GC performance is linear:

<img width="623" alt="Screen Shot 2019-07-23 at 1 30 17 PM" src="https://user-images.githubusercontent.com/169124/61733551-0b4a8c00-ad4e-11e9-98ef-ebacf336298b.png">

[Spreadsheet](https://docs.google.com/spreadsheets/d/1A3hcRjVGHc1UVACM6CoCX6kz6dOkRZqBD_CnHnv_UHQ/edit?usp=sharing)